### PR TITLE
Add support for post generation hooks

### PR DESCRIPTION
**Pull request by gekmihesg**
Alternative approach to #79

### CODE REVIEW

**Comment by Project0** [comment]
I am currently workaround with a pacman hook, but this is not always catching the rebuild of a image. This would solve the issue :+1:

**Review by Molytho — APPROVED** [review]

**Comment by Molytho** [comment]
Maybe there could be a way (e.g when script exits with 255) to notify mkinitcpio about a fatal error so that for example when signing fails the build process fails with a "fatal" instead of only a "warning"

**Comment by gekmihesg** [comment]
Should we just assume that all errors are fatal? A warning could also be printed by the hook executable itself.
Either way, what should happen on a fatal error? Stop processing other hooks and make mkinitpio exit 1?

**Comment by Molytho** [comment]
I'm fine with just always assuming a fatal error.
Exiting with 1 and stop processing is what I thought of.

I just think that the user should get a proper feedback for error when a hook is crucial for the specific setup (like secureboot signing).
Also (somewhere in the future) there might be hooks that depend on other hooks to be run in advance and just continuing after a hook failed might break there assumptions.

**Comment by gekmihesg** [comment]
Agreed, fixed

File: mkinitcpio [hunk 1]
```diff
@@ -458,6 +460,37 @@ preload_builtin_modules() {
...
+    local args=("$@")
+    local hook status
+    local -i count=0
+    local -a paths seen
+
+    IFS=: read -ra paths <<<"$_d_post"
+
+    for path in "${paths[@]}"; do
+        for hook in "$path"/*; do
+            [[ -e $hook ]] && ! in_array "${hook##*/}" "${seen[@]}" || continue
```
Comment (IsaacG) [inline mkinitcpio:473]:
```suggestion
            if [[ -e "$hook" ]] && ! in_array "${hook##*/}" "${seen[@]}"; then
                continue
            fi
```

File: mkinitcpio [hunk 2]
```diff
@@ -458,6 +460,37 @@ preload_builtin_modules() {
...
+            seen+=("${hook##*/}")
+
+            [[ -x $hook ]] || continue
```
Comment (IsaacG) [inline mkinitcpio:476]:
```suggestion
            [[ -x "$hook" ]] || continue
```

File: mkinitcpio [hunk 3]
```diff
@@ -458,6 +460,37 @@ preload_builtin_modules() {
...
+            status=$?
+
+            if (( status )); then
+                error "'%s' failed with exit code %d" "$hook" "$status"
+                return 1
+            fi
+        done
+    done
+
+    ! (( count )) || msg "Post processing done"
```
Comment (IsaacG) [inline mkinitcpio:491]:
```suggestion
    (( count )) && msg "Post processing done"
```

File: mkinitcpio [hunk 4]
```diff
@@ -608,9 +642,10 @@ if [[ -t 1 ]] && (( _optcolor )); then
     try_enable_color
 fi
 
-if [[ -n $_d_flag_hooks && -n $_d_flag_install ]]; then
+if [[ -n $_d_flag_hooks && -n $_d_flag_install && -n $_d_flag_post ]]; then
```
Comment (IsaacG) [inline mkinitcpio:645]:
```suggestion
if [[ -n "$_d_flag_hooks" && -n "$_d_flag_install" && -n "$_d_flag_post" ]]; then
```

File: mkinitcpio [hunk 5]
```diff
@@ -714,15 +749,21 @@ ldconfig -r "$BUILDROOT" &>/dev/null
 umask 077
 
 if [[ $_optgenimg ]]; then
-    build_image "$_optgenimg" "$_optcompress"
+    build_image "$_optgenimg" "$_optcompress" &&
+        _generated+=("$_optgenimg")
```
Comment (IsaacG) [inline mkinitcpio:753]:
```suggestion
    build_image "$_optgenimg" "$_optcompress" && _generated+=("$_optgenimg")
```

**Comment by klausenbusk** [comment]
Moved to to: https://gitlab.archlinux.org/archlinux/mkinitcpio/mkinitcpio/-/merge_requests/81.

**Pull request closed without merging**